### DISCUSSION TITLE
Maximization with a user-defined function

### DISCUSSION THREAD

**Discussion by sabrivaran** [kernc/backtesting.py#310] [Q&A]
Hello, I am working on  a simple ema crossover strategy in a currency USDTRY and I come up with the "problem" that:
    When optimizing my strategy with maximize='Equity Final [$]'  ; equity is calculated in TRY not in USD. 
    Hence I need your help for defining an optimization function which will be based on the value Equity/ data.Close, so that I think it will be in USD. Defining such a function is easy probably but I could not figure it out how it will work with the strategy.
thanks from now.
  I am definitely new to Python and Backtesting so any idea, help or recommendation is really valuable for me
Here is my code:
```py
from backtesting import Backtest, Strategy
from backtesting.lib import crossover, barssince
import pandas as pd
import tulipy as tu 

datafile = """....\\USD_TRYinvesting 01.11.2019-15.04.2021.csv"""
data=pd.read_csv(datafile, index_col = 'Tarih',decimal=',')

data.index = pd.to_datetime(data.index, dayfirst=True)
data.columns = ["Close","Open","High","Low","Volume"]             
data=data.sort_index()
class SmaCross(Strategy):

    ema_no1=9
    ema_no2=20
    
    def init(self):        
        price = self.data.Close
        self.ema1 = self.I(tu.ema, price, self.ema_no1)
        self.ema2 = self.I(tu.ema, price, self.ema_no2)
        
    def next(self):
        if barssince(self.data.High ==5.7525 and self.data.Close==5.7124) > self.ema_no2:       
            if crossover(self.ema1, self.ema2):
               # self.position.close()
                self.buy()
            elif crossover(self.ema2, self.ema1):
               # self.position.close()
                self.sell()


bt = Backtest(data, SmaCross, commission=0.0004,cash=1000,
                exclusive_orders=True)

stats, heatmap = bt.optimize(ema_no1=range(5,15, 1),
                    ema_no2=range(10,50,1),
                    maximize='Equity Final [$]',
                    constraint=lambda param: param.ema_no1 < param.ema_no2,
                    return_heatmap=True)

bt.plot()
mytrades=stats._trades
mytrades.to_csv("trades3.csv")

hm = heatmap.groupby(['ema_no1', 'ema_no2']).mean().unstack()
from backtesting.lib import plot_heatmaps
plot_heatmaps(heatmap, agg='mean')
```

**Comment by kernc** [comment]
[`maximize=` takes a function](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Backtest.optimize). Try:
```py
final_price_in_try = data.Close.iloc[-1]

stats, heatmap = bt.optimize(..., maximize=lambda stats: stats['Equity Final [$]'] / final_price_in_try)
```

**Reply by sabrivaran** [reply]
Thank you for your quick reply but unfortunately that is not giving the result I am looking for. It seems to me that with your maximization function Equity variable is just divided by the dataset's last closing price.

What I need is; dividing Equity during the strategy by the every bar's close price

**Comment by kernc** [comment]
> my strategy with maximize='Equity Final [$]' ; equity is calculated in TRY not in USD

That's what USDTRY currency _is_. If you want to trade the inverse (and have stats based in USD), just feed the strategy with:
```py
data[['Open', 'High', 'Low', 'Close']] = 1 / data[['Open', 'High', 'Low', 'Close']]
```
